File: LeetCode Problems/201 to 300/279. Perfect Squares/279. Perfect Squares.cpp
class Solution {
    vector<int> dp;
    const int inf = 10e4+7;
public:
    int numSquares(int n) {     
        dp.assign(n + 1, inf);
        dp[0] = 0;
        for(int i = 1; i <= n; i++) {
            for(int j = 0; j * j <= i; j++)
                dp[i] = min(dp[i], 1 + dp[i - j * j]);
        }        
        return dp[n];
    }    
};

File: LeetCode Problems/201 to 300/279. Perfect Squares/279. Perfect Squares.java
class Solution {
    public int numSquares(int n) {
        ArrayList<Integer> dp = new ArrayList<>();
        dp.add(0);
        for (int i = 1; i <= n; i++) dp.add(100007);
        for (int i = 1; i <= n; i++) {
            for (int j = 0; j * j <= i; j++) dp.set(i, Math.min(dp.get(i), 1 + dp.get(i - j * j)));
        }
        return dp.get(n);
    }
}
